Update ocaml on Windows: add opam 2.2

RadioPotin opened the pull request:
Greetings,

@rjbou and myself have updated [OCaml on Windows](https://ocaml.org/docs/ocaml-on-windows) tutorial to add opam workflows as the recommended one for installing OCaml on Windows.

Best regards,

RadioPotin committed:
Add opam for Windows installation tutorial

Co-authored-by: Raja Boujbel <raja.boujbel@ocamlpro.com>

cuihtlauac commented:
Beautiful! Thank you very much, @RadioPotin and @rjbou, this is awesome. 
@Sudha247, @dra27, @jonahbeckford: any views on this?

Sudha247 commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,211 @@ description: >
+One detail to address about having a fully functional opam installation is
+related to Git.
+
+On Windows, there are many ways to have a functioning Git installation. Opam will
```
Isn't it recommended to obtain git via `winget`, too?

rjbou replied:
It is indeed. This sentence purpose is to introduce the Git opam init section.

Sudha247 commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -92,6 +286,8 @@ easier way to get a working Windows environment on your machine.
 
 ### Visual Studio Code on Windows
 
+**If you use opam installation**, you will need to add opam switch prefix on your path that runs VSCode.
```
Perhaps there could be a link here on how to add switch prefix to the path?

jonahbeckford replied:
Why is that needed at all? If it is needed, we should file a bug somewhere since that is quite unpleasant for a user.

rjbou replied:
I'll ping @dra27 on this one

Sudha247 reviewed (commented):
Wonderful, this reads very nicely! Huge +1 from my side. Some minor comments below.

jonahbeckford commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,211 @@ description: >
+
+## DkML
+
+There is also [DkML] Windows installer. However, while [DkML] has a modern OCaml 4.14.0 compiler,
```
nit: 4.14.2

rjbou replied:
```suggestion
There is also [DkML] Windows installer. However, while [DkML] has a modern OCaml 4.14.2 compiler,
```

jonahbeckford commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,211 @@ description: >
+In order to have access to OCaml on a Windows PC, users have to use opam so that's our
+first step.
+
+Windows Developers out there will already be familiar with `winget` which is
```
I wouldn't be surprised if less than half of devs know about winget. Even less actively using it. Perhaps I'm wrong, but I think it would be better to just remove the claim.

jonahbeckford commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,211 @@ description: >
-installer that we recommend for new users. However, while [DkML] has a modern OCaml 4.14.0 compiler,
+## opam
+
+Opam now features a fully native Windows compatible installation process that we
```
I get the sentiment, but it isn't accurate to describe a Cygwin installation process as a fully native Windows compatible installation process. IMHO the real win is how easy it is to install Windows with opam 2.2 ... very few questions asked, no prereqs other than Git, etc. ... and the ease of install could be a good lead sentence.

jonahbeckford commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
````diff
@@ -6,8 +6,211 @@ description: >
+Windows Developers out there will already be familiar with `winget` which is
+the default package installer for Windows. Opam is now distributed with it. 
+
+```shell-session
````
Thanks for showing the output! Makes it easy for user to follow and also shows them it is not complicated.

jonahbeckford commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,211 @@ description: >
+sandboxed environment.
+
+You will notice that the repository information fetching stage takes a while to
+complete. This is normal and related to how files are handled on Windows, we
```
It is a bit weird to be assigning blame in a tutorial, but especially blaming Windows to Windows users is weirder still. (Also a bit inaccurate ... I would say / have said that the root cause is thinking that littering files on a filesystem is an adequate replacement for databases and package indexes, a line of thinking that most other package managers reject. Akin to blaming the performance without considering the data structure. But the broader point is all opinions are argumentative with no net benefit)

jonahbeckford commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
````diff
@@ -6,8 +6,211 @@ description: >
+fully functional OCaml compiler available to you:
+
+```shell-session
+> for /f "tokens=*" %i in ('opam env --switch=default') do @%i
````
Correct me if  I'm wrong, but if you install winget in a terminal, and then install opam using winget, any subsequent opam like `opam env` will fail with Not Found, right? Because opam won't be in the PATH.  If so, I think we need to insert a step to close and reopen the terminal.

rjbou replied:
True, this is specified at the [beginning](https://github.com/ocaml/ocaml.org/pull/2727/files#diff-30ebdb1b4b6fb66d5acb62775dcd5f2481c30c431b2e4a304b345359fbf19e86R45):
> Once the package is installed, you can launch a new shell to have access to
your fresh opam binary.

Here, it is have access to OCaml after opam init, so we already have a terminal that have an opam binary in the path.

tmattio commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -38,15 +241,6 @@ The guidance is based on the availability table below:
-
 ## Installation Environments
 
 ### `opam-repository-mingw`
```
I think we can remove this section entirely since it is deprecated and opam-repository is now compatible with Windows.

christinerose committed:
Minor line editing / formatting on 3_01_ocaml_on_windows.md

rjbou commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,211 @@ description: >
+first step.
+
+Windows Developers out there will already be familiar with `winget` which is
+the default package installer for Windows. Opam is now distributed with it. 
```
```suggestion
Opam is now distributed via `winget` for Windows.
```

rjbou commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,211 @@ description: >
+You will notice that the repository information fetching stage takes a while to
+complete. This is normal and related to how files are handled on Windows, we
+advice our users to go get themselves their favourite hot beverage while it
+runs.
```
```suggestion
You will notice that the repository information fetching stage takes a while to
complete. This is normal (for the moment), we advice our users to go get 
themselves their favourite hot beverage while it runs.
```

rjbou reviewed (commented):
Thanks all for the reviews! I've addressed some in comments.

RadioPotin committed:
edit section mentioning winget

Co-authored-by: R. Boujbel <rjbou@ocamlpro.com>

RadioPotin committed:
remove windows file management comment

Co-authored-by: R. Boujbel <rjbou@ocamlpro.com>

RadioPotin committed:
Change OCaml compiler version to 4.14.2

Co-authored-by: R. Boujbel <rjbou@ocamlpro.com>

RadioPotin commented:
Hey again everybody, 

Thank you ever so much for the constructive suggestions, hopefully this version is now mergeable,

Kind regards,

qrpnxz commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,8 +6,209 @@ description: >
+
+There is also [DkML] Windows installer. However, while [DkML] has a modern OCaml 4.14.2 compiler,
 it does not track the latest OCaml compilers. We will officially support Windows as a Tier 1
 platform with a [major release of opam](#opam-22) in the coming months, and it will be compatible with
```
> `We `**`will`**` officially support Windows as a Tier 1 platform with a [major release of opam]`**`(#opam-22) in the coming months,`**

Already happened, hence this PR. This link will be broken.

qrpnxz commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -38,21 +239,12 @@ The guidance is based on the availability table below:
 ╰──────────────────────────────────────────────────────────────────────────────────────────╯
```
> `│ Tier 1 │ OCaml 5 with Opam 2.2             │ Full support. `**`Coming in the next few months`**` │`

Already happened.

qrpnxz commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
````diff
@@ -6,8 +6,209 @@ description: >
+```shell-session
+<><> Git ><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
+Cygwin Git is functional but can have credentials issues for private repositories, we recommend using:
+  - Install via 'winget install Git.Git'
````
This would be good to know from the outset. https://ocaml.org/install#windows recommends doing
```
winget install Git.Git OCaml.opam
```

qrpnxz reviewed (changes requested):
A couple of details still make reference to a future release of opam 2.2.
Installation of git at the same time as opam should be suggested.

christinerose committed:
Removed unnecessary work and fixed spelling error

tmattio committed:
Incorporate changes from #554

Co-authored-by: Jonah Beckford <9566106-jonahbeckford@users.noreply.gitlab.com>

tmattio committed:
Incorporate PR review and remove stdout

- Change wording of the introduction
- Recommend using winget install Git.Git OCaml.opam
- Shorten instructions when they can be made more concise

tmattio commented:
I've pushed a few changes:
- Integrated changes from https://github.com/ocaml/ocaml.org/pull/554
- Reworded the introduction as suggested by @jonahbeckford 
- Updated the availability table (thanks @qrpnxz)
- Updated the recommended instructions to be `winget install Git.Git OCaml.opam` (thanks @qrpnxz)
- Removed the stdout from the shell blocks, this was making the document very verbose for not much value I think
- Shortened some instructions when they could be made more concise.

I think this is ready to go, but I'll wait for a thumbs up from @RadioPotin @rjbou if the changes above look good to you, and give a couple days for a last round of review.

rjbou commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,73 +6,105 @@ description: >
 
-## Installation Environments
+opam requires a Unix-like environment to function. By default, opam relies on `cygwin` and is also compatible with
+`msys2`.
```
```suggestion
opam requires a Unix-like environment to function. By default, 
opam relies on Cygwin and is also compatible with MSYS2.
```

rjbou commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,73 +6,105 @@ description: >
+letting it create its own internal Cygwin installation that will remain managed
+by opam. This cuts down the possible interferences that other tools
+interacting with such environments might introduce. Think of it as a
+sandboxed environment.
```
```suggestion
```
It's duplicated

rjbou commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
```diff
@@ -6,73 +6,105 @@ description: >
+Opam's default behavior when initializing is to install a fresh `switch` as
+well as an OCaml compiler of version `> 4.05`. By default, opam chooses `mingw` as
+a C compiler when creating switches, but know that you can choose an alternative to it
+instead, like `msvc` with the following command:
```
```suggestion
a C compiler when creating switches, but know that you can specify to install an 
alternative to it instead, like `msvc`, with the following command:
```

rjbou commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
````diff
@@ -6,73 +6,105 @@ description: >
+
+After `opam init` completes, run the following command to update your environment:
+
+```
+> for /f "tokens=*" %i in ('opam env --switch=default') do @%i
+```
````
That command is only for CMD
````suggestion
On CMD:
```
> for /f "tokens=*" %i in ('opam env --switch=default') do @%i
```
On PowerShell:
```
> (& opam env --switch=default) -split '\\r?\\n' | ForEach-Object { Invoke-Expression $_ }
```
````

rjbou commented on data/tutorials/getting-started/3_01_ocaml_on_windows.md:
````diff
@@ -6,73 +6,105 @@ description: >
+```
+> for /f "tokens=*" %i in ('opam env --switch=default') do @%i
+```
+
````
Users may think that they need to keep that command somewhere to find easily.
```suggestion
Opam will display the shell update command each time it is needed.

```

rjbou reviewed (commented):
Thanks all for these inputs!
I'm not sure for the removal of shell output: as pointed out Jonah, it helps users to know should happen, but if the rest of the website is without shell output, it's better to remove indeed. 
Remain tiny comments and we should be good to go.

tmattio committed:
Apply suggestions from code review

Co-authored-by: R. Boujbel <rjbou@ocamlpro.com>

tmattio committed:
Update data/tutorials/getting-started/3_01_ocaml_on_windows.md

Co-authored-by: R. Boujbel <rjbou@ocamlpro.com>

tmattio commented:
Thanks @rjbou! All suggestions applied. We can bring back the shell output if others think this contributes to the clarity of the document, but indeed, we don't do that in the rest of the documentation.

christinerose committed:
line edits 3_01_ocaml_on_windows.md

tmattio commented:
Merging now as there isn't any objection, follow up PRs to improve on it are most welcome though!

Thanks everyone for the contribution and reviews!
